mip/driver_imx_rt1020.c

```diff
@@ -0,0 +1,302 @@
+    tx_buffer_descriptor[i].control = ((i<(ENET_RXBD_NUM-1))?0:BIT_SET(13)) | BIT_SET(10); // (W*)+TC
+    tx_buffer_descriptor[i].buffer = (uint32_t *)tx_data_buffer[i];
+  }
+
+  // Continue ENET configuration
+  ENET->RDSR = (uint32_t)(uintptr_t)rx_buffer_descriptor;
+  ENET->TDSR = (uint32_t)(uintptr_t)tx_buffer_descriptor;
+  ENET->MRBR[0] = ENET_RXBUFF_SIZE; // Same size for RX/TX buffers
+
+  // MAC address filtering (bytes in reversed order)
+  ENET->PAUR = ((uint32_t) mac[4] << 24U) | (uint32_t) mac[5] << 16U;
+  ENET->PALR = (uint32_t) (mac[0] << 24U) | ((uint32_t) mac[1] << 16U) |
+                 ((uint32_t) mac[2] << 8U) | mac[3];
+
+  // Init Hash tables (mac filtering)
+  ENET->IAUR = 0; // Unicast
+  ENET->IALR = 0;
+  ENET->GAUR = 0; // Multicast
+  ENET->GALR = 0;
+
+  // Set ENET Online
+  ENET->ECR |= BIT_SET(8); // ENET Set Little-endian + (FEC buffer desc.)
+  ENET->ECR |= BIT_SET(1); // Enable
+
+  // Set interrupt mask
+  ENET->EIMR = EIMR_RX_ERR;
+
+  // RX Descriptor activation
+  ENET->RDAR = BIT_SET(24); // Activate Receive Descriptor
+
+  (void) data;
+  return true;
+}
+
+// Transmit frame
+static uint32_t s_txno;
+
+static size_t mip_driver_imx_rt1020_tx(const void *buf, size_t len, void *userdata) {
+
+  if (len > sizeof(tx_data_buffer[ENET_TXBD_NUM])) {
+  //  MG_ERROR(("Frame too big, %ld", (long) len));
+    len = 0;  // Frame is too big
+  } else if ((tx_buffer_descriptor[s_txno].control & BIT_SET(15))) {
+  MG_ERROR(("No free descriptors"));
+    // printf("D0 %lx SR %lx\n", (long) s_txdesc[0][0], (long) ETH->DMASR);
+    len = 0;  // All descriptors are busy, fail
+  } else {
+    memcpy(tx_data_buffer[s_txno], buf, len);     // Copy data
+    tx_buffer_descriptor[s_txno].length = (uint16_t) len;  // Set data len
+    tx_buffer_descriptor[s_txno].control |= (uint16_t)(BIT_SET(10)); // TC (transmit CRC)
+    //  tx_buffer_descriptor[s_txno].control &= (uint16_t)(BIT_SET(14) | BIT_SET(12)); // Own doesn't affect HW
+    tx_buffer_descriptor[s_txno].control |= (uint16_t)(BIT_SET(15) | BIT_SET(11)); // R+L (ready+last)
+    ENET->TDAR = BIT_SET(24); // Descriptor updated. Hand over to DMA.
+    // INFO
+    // Relevant Descriptor bits: 15(R)  Ready
+    //                           11(L)  last in frame
+    //                           10(TC) transmis CRC
+    // __DSB(); // ARM errata 838869 Cortex-M4, M4F, M7, M7F: "store immediate overlapping
+                // exception" return might vector to incorrect interrupt.
+    if (++s_txno >= ENET_TXBD_NUM) s_txno = 0;
+  }
+  (void) userdata;
+  return len;
+}
+
+// IRQ (RX)
+static uint32_t s_rxno;
+
+void ENET_IRQHandler(void) {
+  ENET->EIMR = 0;           // Mask interrupts.
+  uint32_t eir = ENET->EIR; // Read EIR
+  ENET->EIR = 0xffffffff;   // Clear interrupts
+
+  if (eir & EIMR_RX_ERR) // Global mask used
+  {
+    if (rx_buffer_descriptor[s_rxno].control & BIT_SET(15)) {
+      ENET->EIMR = EIMR_RX_ERR; // Enable interrupts
```

Comment:
I personally don't like fiddling with enabling and disabling interrupts, particularly within their own handler. AFAIK IRQs in ARM are not reentrant, there is an _active_ flag in the NVIC.

Reply:
Ok, i need to check this further. The IRQ goes back a few function calls away. It has not to be re-entrant.